bug: unexpected panic

version: `v0.19.2`

```
2025-09-05 09:59:53.685	
  90:                0x0 - <unknown>

2025-09-05 09:59:53.685	
                               at ./misc/../sysdeps/unix/sysv/linux/aarch64/clone3.S:76:0
2025-09-05 09:59:53.685	
  89:     0xffff9f0cba4c - thread_start
2025-09-05 09:59:53.685	
                               at ./nptl/pthread_create.c:447:8
2025-09-05 09:59:53.685	
  88:     0xffff9f06595c - start_thread
2025-09-05 09:59:53.685	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/sys/pal/unix/thread.rs:109:17
2025-09-05 09:59:53.685	
  87:     0xaaaaf0ea70ac - std::sys::pal::unix::thread::Thread::new::thread_start::h3cd332fa8e79d294
2025-09-05 09:59:53.685	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/alloc/src/boxed.rs:1972:9
2025-09-05 09:59:53.685	
  86:     0xaaaaf0ea70ac - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hff1fa87650c6a275
2025-09-05 09:59:53.685	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/alloc/src/boxed.rs:1972:9
2025-09-05 09:59:53.685	
  85:     0xaaaaf0ea70ac - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hb062ada926193a82
2025-09-05 09:59:53.685	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/core/src/ops/function.rs:250:5
2025-09-05 09:59:53.685	
  84:     0xaaaaf0f65a48 - core::ops::function::FnOnce::call_once{{vtable.shim}}::h2a4ff37f8e7f82a9
2025-09-05 09:59:53.685	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/thread/mod.rs:557:30
2025-09-05 09:59:53.685	
  83:     0xaaaaf0f65a48 - std::thread::Builder::spawn_unchecked_::{{closure}}::h6192aee9fb3ec568
2025-09-05 09:59:53.685	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/panic.rs:359:14
2025-09-05 09:59:53.685	
  82:     0xaaaaf0f65a48 - std::panic::catch_unwind::hff04af3045cad4b2
2025-09-05 09:59:53.685	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/panicking.rs:550:19
2025-09-05 09:59:53.685	
  81:     0xaaaaf0f65a48 - std::panicking::try::hc6e9c7c96f86c684
2025-09-05 09:59:53.685	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/panicking.rs:587:40
2025-09-05 09:59:53.685	
  80:     0xaaaaf0f65a48 - std::panicking::try::do_call::h0dd81d705b9efafe
2025-09-05 09:59:53.685	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/core/src/panic/unwind_safe.rs:272:9
2025-09-05 09:59:53.685	
  79:     0xaaaaf0f65a48 - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::hc44977284c350e16
2025-09-05 09:59:53.685	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/thread/mod.rs:559:17
2025-09-05 09:59:53.685	
  78:     0xaaaaf0f65a48 - std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}::h6e46aaf15a1bfbc9
2025-09-05 09:59:53.685	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/sys/backtrace.rs:152:18
2025-09-05 09:59:53.685	
  77:     0xaaaaf0f63e08 - std::sys::backtrace::__rust_begin_short_backtrace::he0b38cedea9e0ad6
2025-09-05 09:59:53.685	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/blocking/pool.rs:469:13
2025-09-05 09:59:53.685	
  76:     0xaaaaf0f63e08 - tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}::h1726e5aafcad36fc
2025-09-05 09:59:53.685	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/blocking/pool.rs:511:17
2025-09-05 09:59:53.685	
  75:     0xaaaaf0f2b914 - tokio::runtime::blocking::pool::Inner::run::h0013718426cf074d
2025-09-05 09:59:53.685	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/blocking/pool.rs:161:9
2025-09-05 09:59:53.685	
  74:     0xaaaaf0f2b914 - tokio::runtime::blocking::pool::Task::run::h2218ef1e0d63c090
2025-09-05 09:59:53.685	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/task/mod.rs:500:9
2025-09-05 09:59:53.685	
  73:     0xaaaaf0f2b914 - tokio::runtime::task::UnownedTask<S>::run::h9c6914f98caf36f6
2025-09-05 09:59:53.685	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/task/raw.rs:201:18
2025-09-05 09:59:53.685	
  72:     0xaaaaf0f2b914 - tokio::runtime::task::raw::RawTask::poll::h6dc24478fd17b210
2025-09-05 09:59:53.685	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/task/harness.rs:154:15
2025-09-05 09:59:53.685	
  71:     0xaaaaf0f221f0 - tokio::runtime::task::harness::Harness<T,S>::poll::h43e4e7cce82343fd
2025-09-05 09:59:53.685	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/task/harness.rs:209:27
2025-09-05 09:59:53.685	
  70:     0xaaaaf0f221f0 - tokio::runtime::task::harness::Harness<T,S>::poll_inner::h090cedad0f309947
2025-09-05 09:59:53.685	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/task/harness.rs:520:18
2025-09-05 09:59:53.685	
  69:     0xaaaaf0f221f0 - tokio::runtime::task::harness::poll_future::hc7295ec2830e8172
2025-09-05 09:59:53.685	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/panic.rs:359:14
2025-09-05 09:59:53.685	
  68:     0xaaaaf0f221f0 - std::panic::catch_unwind::h7c215643a8f75deb
2025-09-05 09:59:53.685	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/panicking.rs:550:19
2025-09-05 09:59:53.685	
  67:     0xaaaaf0f221f0 - std::panicking::try::hb0d4deca205875fa
2025-09-05 09:59:53.685	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/panicking.rs:587:40
2025-09-05 09:59:53.685	
  66:     0xaaaaf0f221f0 - std::panicking::try::do_call::h34fcb196a9bf0d27
2025-09-05 09:59:53.685	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/core/src/panic/unwind_safe.rs:272:9
2025-09-05 09:59:53.685	
  65:     0xaaaaf0f221f0 - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::had1038596e40007e
2025-09-05 09:59:53.685	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/task/harness.rs:532:19
2025-09-05 09:59:53.685	
  64:     0xaaaaf0f221f0 - tokio::runtime::task::harness::poll_future::{{closure}}::h9ecee042e6acfdcd
2025-09-05 09:59:53.685	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/task/core.rs:320:30
2025-09-05 09:59:53.685	
  63:     0xaaaaf0f55500 - tokio::runtime::task::core::Core<T,S>::poll::h74356b9a0b95f933
2025-09-05 09:59:53.685	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/loom/std/unsafe_cell.rs:16:9
2025-09-05 09:59:53.685	
  62:     0xaaaaf0f55500 - tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut::h62dbec162c4cb3d4
2025-09-05 09:59:53.685	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/task/core.rs:331:17
2025-09-05 09:59:53.685	
  61:     0xaaaaf0f55500 - tokio::runtime::task::core::Core<T,S>::poll::{{closure}}::hd3f30fd96ea30926
2025-09-05 09:59:53.685	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tracing-0.1.41/src/instrument.rs:321:9
2025-09-05 09:59:53.685	
  60:     0xaaaaf0f33f1c - <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll::hb29b76cf241b2435
2025-09-05 09:59:53.685	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/blocking/task.rs:42:21
2025-09-05 09:59:53.685	
  59:     0xaaaaf0f33f1c - <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll::h12623ed06b19d352
2025-09-05 09:59:53.685	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/scheduler/multi_thread/worker.rs:460:45
2025-09-05 09:59:53.685	
  58:     0xaaaaf0f33f1c - tokio::runtime::scheduler::multi_thread::worker::Launch::launch::{{closure}}::h7667e6940aa3ea28
2025-09-05 09:59:53.685	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/scheduler/multi_thread/worker.rs:494:5
2025-09-05 09:59:53.684	
  57:     0xaaaaf0f4e364 - tokio::runtime::scheduler::multi_thread::worker::run::hfef1ca8e3ffb5062
2025-09-05 09:59:53.684	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/context/runtime.rs:65:16
2025-09-05 09:59:53.684	
  56:     0xaaaaf0f4e364 - tokio::runtime::context::runtime::enter_runtime::h7737761e27d9b027
2025-09-05 09:59:53.684	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/scheduler/multi_thread/worker.rs:502:9
2025-09-05 09:59:53.684	
  55:     0xaaaaf0f4e364 - tokio::runtime::scheduler::multi_thread::worker::run::{{closure}}::hf3947c6de66ebb86
2025-09-05 09:59:53.684	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/context.rs:180:9
2025-09-05 09:59:53.684	
  54:     0xaaaaf0f4e364 - tokio::runtime::context::set_scheduler::hb9e2aa45d8139727
2025-09-05 09:59:53.684	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/thread/local.rs:275:15
2025-09-05 09:59:53.684	
  53:     0xaaaaf0f4e364 - std::thread::local::LocalKey<T>::with::hf6b60c0ca17f9b70
2025-09-05 09:59:53.684	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/thread/local.rs:311:12
2025-09-05 09:59:53.684	
  52:     0xaaaaf0f4e364 - std::thread::local::LocalKey<T>::try_with::h9a58772cec7fa4a3
2025-09-05 09:59:53.684	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/context.rs:180:26
2025-09-05 09:59:53.684	
  51:     0xaaaaf0f4e364 - tokio::runtime::context::set_scheduler::{{closure}}::h7bb4e72d27154f45
2025-09-05 09:59:53.684	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/context/scoped.rs:40:9
2025-09-05 09:59:53.684	
  50:     0xaaaaf0f2fadc - tokio::runtime::context::scoped::Scoped<T>::set::hfabf85351a92cbd5
2025-09-05 09:59:53.684	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/scheduler/multi_thread/worker.rs:507:21
2025-09-05 09:59:53.684	
  49:     0xaaaaf0f2fadc - tokio::runtime::scheduler::multi_thread::worker::run::{{closure}}::{{closure}}::hcd3b6c29d8fd4434
2025-09-05 09:59:53.684	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/scheduler/multi_thread/worker.rs:542:24
2025-09-05 09:59:53.684	
  48:     0xaaaaf0f4ed70 - tokio::runtime::scheduler::multi_thread::worker::Context::run::h765a6f9fd3822d4d
2025-09-05 09:59:53.684	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/scheduler/multi_thread/worker.rs:594:9
2025-09-05 09:59:53.684	
  47:     0xaaaaf0f4fa34 - tokio::runtime::scheduler::multi_thread::worker::Context::run_task::hd82af526d5a6f093
2025-09-05 09:59:53.684	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/task/coop/mod.rs:133:5
2025-09-05 09:59:53.684	
  46:     0xaaaaf0f4fa34 - tokio::task::coop::budget::h7be62ce1bf4ec252
2025-09-05 09:59:53.684	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/task/coop/mod.rs:167:5
2025-09-05 09:59:53.684	
  45:     0xaaaaf0f4fa34 - tokio::task::coop::with_budget::hb1f9f80a450fc130
2025-09-05 09:59:53.684	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/scheduler/multi_thread/worker.rs:674:22
2025-09-05 09:59:53.684	
  44:     0xaaaaf0f4fa34 - tokio::runtime::scheduler::multi_thread::worker::Context::run_task::{{closure}}::he263a7193830efa0
2025-09-05 09:59:53.684	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/task/mod.rs:463:9
2025-09-05 09:59:53.684	
  43:     0xaaaaf0f4fa34 - tokio::runtime::task::LocalNotified<S>::run::h91b1595e70e8be27
2025-09-05 09:59:53.684	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/task/raw.rs:201:18
2025-09-05 09:59:53.684	
  42:     0xaaaaf0f4fa34 - tokio::runtime::task::raw::RawTask::poll::h6dc24478fd17b210
2025-09-05 09:59:53.684	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/task/harness.rs:154:15
2025-09-05 09:59:53.684	
  41:     0xaaaaefe4a158 - tokio::runtime::task::harness::Harness<T,S>::poll::hc2f2583a502ac85d
2025-09-05 09:59:53.684	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/task/harness.rs:209:27
2025-09-05 09:59:53.684	
  40:     0xaaaaefe4a158 - tokio::runtime::task::harness::Harness<T,S>::poll_inner::h600083d77d68bd28
2025-09-05 09:59:53.684	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/task/harness.rs:520:18
2025-09-05 09:59:53.684	
  39:     0xaaaaefe4a158 - tokio::runtime::task::harness::poll_future::hb4670cfe61eabb81
2025-09-05 09:59:53.684	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/panic.rs:359:14
2025-09-05 09:59:53.684	
  38:     0xaaaaefe4a158 - std::panic::catch_unwind::h2655c4fb13ba70ff
2025-09-05 09:59:53.684	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/panicking.rs:550:19
2025-09-05 09:59:53.684	
  37:     0xaaaaefe4a158 - std::panicking::try::h38529c35d950df0b
2025-09-05 09:59:53.684	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/panicking.rs:587:40
2025-09-05 09:59:53.684	
  36:     0xaaaaefe4a158 - std::panicking::try::do_call::h3458c0a24e2f3ea0
2025-09-05 09:59:53.684	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/core/src/panic/unwind_safe.rs:272:9
2025-09-05 09:59:53.684	
  35:     0xaaaaefe4a158 - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::hb6b4ed84e6940b20
2025-09-05 09:59:53.684	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/task/harness.rs:532:19
2025-09-05 09:59:53.684	
  34:     0xaaaaefe4a158 - tokio::runtime::task::harness::poll_future::{{closure}}::h41a483a242b9e4c5
2025-09-05 09:59:53.684	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/task/core.rs:320:30
2025-09-05 09:59:53.684	
  33:     0xaaaaefe4a158 - tokio::runtime::task::core::Core<T,S>::poll::h624edf58825c6412
2025-09-05 09:59:53.684	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/loom/std/unsafe_cell.rs:16:9
2025-09-05 09:59:53.684	
  32:     0xaaaaefe4a158 - tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut::h9429505d71da5c13
2025-09-05 09:59:53.684	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/task/core.rs:331:17
2025-09-05 09:59:53.684	
  31:     0xaaaaefe4a158 - tokio::runtime::task::core::Core<T,S>::poll::{{closure}}::h184fc9cd6df4659f
2025-09-05 09:59:53.684	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tracing-0.1.41/src/instrument.rs:321:9
2025-09-05 09:59:53.684	
  30:     0xaaaaefe4a158 - <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll::h3619296f8d432865
2025-09-05 09:59:53.684	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/fastrace-0.7.8/src/future.rs:124:19
2025-09-05 09:59:53.683	
  29:     0xaaaaefe8f748 - <fastrace::future::InSpan<T> as core::future::future::Future>::poll::h7240cdc7443ef5c1
2025-09-05 09:59:53.683	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/foyer-memory-0.19.2/src/raw.rs:1188:50
2025-09-05 09:59:53.683	
  28:     0xaaaaefe8f748 - foyer_memory::raw::RawCache<E,S,I>::fetch_inner::{{closure}}::heb4be75217d736bd
2025-09-05 09:59:53.683	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/foyer-memory-0.19.2/src/raw.rs:592:9
2025-09-05 09:59:53.683	
  27:     0xaaaaef653810 - foyer_memory::raw::RawCache<E,S,I>::insert_with_properties::hb3382c078a035210
2025-09-05 09:59:53.683	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/foyer-memory-0.19.2/src/raw.rs:629:14
2025-09-05 09:59:53.683	
  26:     0xaaaaef651e08 - foyer_memory::raw::RawCache<E,S,I>::insert_inner::ha932c7f607a68f14
2025-09-05 09:59:53.683	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/foyer-common-0.19.2/src/utils/scope.rs:23:9
2025-09-05 09:59:53.683	
  25:     0xaaaaef651e08 - foyer_common::utils::scope::Scope::with::hd5050a878b269cc9
2025-09-05 09:59:53.683	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/foyer-memory-0.19.2/src/raw.rs:629:31
2025-09-05 09:59:53.683	
  24:     0xaaaaef651e08 - foyer_memory::raw::RawCache<E,S,I>::insert_inner::{{closure}}::ha52df8dd43798694
2025-09-05 09:59:53.683	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/foyer-memory-0.19.2/src/raw.rs:150:9
2025-09-05 09:59:53.683	
  23:     0xaaaaef651e08 - foyer_memory::raw::RawCacheShard<E,S,I>::emplace::h818eb87a92a67e0a
2025-09-05 09:59:53.683	
                               at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/foyer-memory-0.19.2/src/raw.rs:126:72
2025-09-05 09:59:53.683	
  22:     0xaaaaef651e08 - foyer_memory::raw::RawCacheShard<E,S,I>::evict::hb4667bb68f01efc3
2025-09-05 09:59:53.683	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/core/src/option.rs:978:21
2025-09-05 09:59:53.683	
  21:     0xaaaaef651e08 - core::option::Option<T>::unwrap::h4809f1523ef555d0
2025-09-05 09:59:53.648	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/core/src/option.rs:2015:5
2025-09-05 09:59:53.648	
  20:     0xaaaae0a1ed50 - core::option::unwrap_failed::heb889c5c253c67a9
2025-09-05 09:59:53.648	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/core/src/panicking.rs:145:5
2025-09-05 09:59:53.648	
  19:     0xaaaae0a1ee28 - core::panicking::panic::h2e20c09ccb914a33
2025-09-05 09:59:53.648	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/core/src/panicking.rs:75:14
2025-09-05 09:59:53.648	
  18:     0xaaaae0a1edb8 - core::panicking::panic_fmt::hc247d0cd13f82f86
2025-09-05 09:59:53.648	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/panicking.rs:695:5
2025-09-05 09:59:53.648	
  17:     0xaaaaf0ea0d80 - rust_begin_unwind
2025-09-05 09:59:53.648	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/sys/backtrace.rs:168:18
2025-09-05 09:59:53.648	
  16:     0xaaaaf0ea003c - std::sys::backtrace::__rust_end_short_backtrace::h02f1caaf7b7ade93
2025-09-05 09:59:53.648	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/panicking.rs:697:13
2025-09-05 09:59:53.648	
  15:     0xaaaaf0ea1098 - std::panicking::begin_panic_handler::{{closure}}::h7106813b6611b27b
2025-09-05 09:59:53.648	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/panicking.rs:839:13
2025-09-05 09:59:53.648	
  14:     0xaaaaf0ea13e8 - std::panicking::rust_panic_with_hook::hb46e3432064fe42e
2025-09-05 09:59:53.648	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/alloc/src/boxed.rs:1986:9
2025-09-05 09:59:53.648	
  13:     0xaaaaf0ea13e8 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::ha99f8ab057c973aa
2025-09-05 09:59:53.648	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/panicking.rs:241:47
2025-09-05 09:59:53.648	
  12:     0xaaaaedb7e938 - std::panicking::update_hook::{{closure}}::hb29908f7baa60697
2025-09-05 09:59:53.648	
                               at /risingwave/src/utils/runtime/src/panic_hook.rs:25:9
2025-09-05 09:59:53.648	
  11:     0xaaaaedb7e938 - risingwave_rt::panic_hook::set_panic_hook::{{closure}}::h7e234b651bc61922
2025-09-05 09:59:53.648	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/alloc/src/boxed.rs:1986:9
2025-09-05 09:59:53.648	
  10:     0xaaaaedb7e938 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h8bbbc41f152af2ee
2025-09-05 09:59:53.646	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/panicking.rs:325:9
2025-09-05 09:59:53.646	
   9:     0xaaaaf0ea079c - std::panicking::default_hook::h46d1ffcddaa8058a
2025-09-05 09:59:53.646	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/panicking.rs:298:22
2025-09-05 09:59:53.646	
   8:     0xaaaaf0ea0ad8 - std::panicking::default_hook::{{closure}}::hbee1dc37e467167b
2025-09-05 09:59:53.646	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/sys/backtrace.rs:42:9
2025-09-05 09:59:53.646	
   7:     0xaaaaf0e9f9e0 - std::sys::backtrace::BacktraceLock::print::h2061cfb3839b2d3c
2025-09-05 09:59:53.646	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/io/mod.rs:1890:15
2025-09-05 09:59:53.646	
   6:     0xaaaaf0e99c00 - std::io::Write::write_fmt::h7b482adf221f73f5
2025-09-05 09:59:53.646	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/core/src/fmt/mod.rs:1449:21
2025-09-05 09:59:53.646	
   5:     0xaaaae0a220a4 - core::fmt::write::h86ac288c85fd4a0a
2025-09-05 09:59:53.646	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/core/src/fmt/rt.rs:177:76
2025-09-05 09:59:53.646	
   4:     0xaaaae0a220a4 - core::fmt::rt::Argument::fmt::h9058bde717e94412
2025-09-05 09:59:53.645	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/sys/backtrace.rs:39:26
2025-09-05 09:59:53.645	
   3:     0xaaaaf0e9fb40 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::hff8abd39e1f90e87
2025-09-05 09:59:53.645	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/sys/backtrace.rs:66:9
2025-09-05 09:59:53.645	
   2:     0xaaaaf0e9fb40 - std::sys::backtrace::_print_fmt::h18ce864b39d76b45
2025-09-05 09:59:53.645	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/../../backtrace/src/backtrace/mod.rs:66:14
2025-09-05 09:59:53.645	
   1:     0xaaaaf0e9fb40 - std::backtrace_rs::backtrace::trace_unsynchronized::h27df4f6f457f4fd5
2025-09-05 09:59:53.645	
                               at /rustc/f9e0239a7bc813b4aceffc7f069f4797cde3175c/library/std/src/../../backtrace/src/backtrace/libunwind.rs:117:9
2025-09-05 09:59:53.645	
   0:     0xaaaaf0e9fb40 - std::backtrace_rs::backtrace::libunwind::trace::h68fa4f2ef0c30713
2025-09-05 09:59:53.645	
stack backtrace:
2025-09-05 09:59:53.645	
called `Option::unwrap()` on a `None` value
2025-09-05 09:59:53.645	
thread 'rw-main' panicked at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/foyer-memory-0.19.2/src/raw.rs:126:72:
```

Maybe related:

https://github.com/foyer-rs/foyer/actions/runs/17758759361/job/50466592206

```
=================================================================
==10022==ERROR: AddressSanitizer: heap-use-after-free on address 0x5250000bf000 at pc 0x558a428bfb5e bp 0x7f551b5a7340 sp 0x7f551b5a6b08
READ of size 4096 at 0x5250000bf000 thread T19
2025-09-16T07:58:26.296970Z  INFO foyer::hybrid::cache: [hybrid]: close consumes 453.425µs
2025-09-16T07:58:26.297563Z  INFO foyer_storage::store: [store]: Dedicated runtime is disabled. This may lead to spikes in latency under high load. Hint: Consider configuring a dedicated runtime.
    #0 0x558a428bfb5d  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x60bb5d) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #1 0x558a44134f2c  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1e80f2c) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #2 0x558a43c3c758  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1988758) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #3 0x558a43c07429  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1953429) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #4 0x558a43c70100  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x19bc100) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #5 0x558a43c94716  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x19e0716) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #6 0x558a43c93447  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x19df447) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #7 0x558a43c61258  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x19ad258) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #8 0x558a43c80bab  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x19ccbab) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #9 0x558a43c2bffc  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1977ffc) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #10 0x558a43c4abea  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1996bea) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #11 0x558a43c3f6d6  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x198b6d6) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #12 0x558a43c5edb8  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x19aadb8) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #13 0x558a43c642d4  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x19b02d4) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #14 0x558a43c6826c  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x19b426c) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #15 0x558a43c1289c  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x195e89c) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #16 0x558a43f9d9cc  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1ce99cc) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #17 0x558a43f0bd35  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1c57d35) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #18 0x558a43f1ea86  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1c6aa86) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #19 0x558a43f236b1  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1c6f6b1) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #20 0x558a43f22ff2  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1c6eff2) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #21 0x558a43fa1e15  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1cede15) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #22 0x558a43f9477d  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1ce077d) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #23 0x558a43fcf090  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1d1b090) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #24 0x558a43f94daf  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1ce0daf) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #25 0x558a43fa0e6a  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1cece6a) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #26 0x558a43f94385  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1ce0385) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #27 0x558a43f43a2e  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1c8fa2e) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #28 0x558a4413f8da  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1e8b8da) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #29 0x558a4292c176  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x678176) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #30 0x7f5525e9caa3  (/lib/x86_64-linux-gnu/libc.so.6+0x9caa3) (BuildId: 282c2c16e7b6600b0b22ea0c99010d2795752b5f)
    #31 0x7f5525f29c3b  (/lib/x86_64-linux-gnu/libc.so.6+0x129c3b) (BuildId: 282c2c16e7b6600b0b22ea0c99010d2795752b5f)

0x5250000bf000 is located 0 bytes inside of 4096-byte region [0x5250000bf000,0x5250000c0000)
freed by thread T2 here:
    #0 0x558a4292e456  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x67a456) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #1 0x558a43c0b494  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1957494) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #2 0x558a43c25bee  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1971bee) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #3 0x558a43c78c8a  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x19c4c8a) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #4 0x558a42ceef5c  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0xa3af5c) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #5 0x558a42986323  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x6d2323) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #6 0x558a43339efd  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1085efd) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #7 0x558a42982300  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x6ce300) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #8 0x558a42be5f1a  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x931f1a) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #9 0x558a4380fe4a  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x155be4a) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #10 0x558a43f0b9b1  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1c579b1) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #11 0x558a43f16fca  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1c62fca) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #12 0x558a43f16dbe  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1c62dbe) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #13 0x558a43f1aec9  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1c66ec9) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #14 0x558a43f137bb  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1c5f7bb) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #15 0x558a43f571b9  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1ca31b9) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #16 0x558a43f5682a  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1ca282a) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #17 0x558a43f1aa84  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1c66a84) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #18 0x558a43f167c6  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1c627c6) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #19 0x558a43f604ca  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1cac4ca) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #20 0x558a43f49046  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1c95046) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #21 0x558a43030b76  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0xd7cb76) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #22 0x558a4386cbc7  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x15b8bc7) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #23 0x558a43830eb3  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x157ceb3) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #24 0x558a43834601  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1580601) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #25 0x558a4413f8da  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1e8b8da) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #26 0x558a4292c176  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x678176) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)

previously allocated by thread T2 here:
    #0 0x558a4292f1b7  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x67b1b7) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #1 0x558a4413902d  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1e8502d) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #2 0x558a43c0af6f  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1956f6f) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #3 0x558a43c25460  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1971460) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #4 0x558a43c257ba  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x19717ba) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #5 0x558a43c2a34d  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x197634d) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #6 0x558a43bf1961  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x193d961) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #7 0x558a43bedaa1  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1939aa1) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #8 0x558a43390c96  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x10dcc96) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #9 0x558a433a0a1c  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x10eca1c) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #10 0x558a42d62922  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0xaae922) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #11 0x558a42d1bb3e  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0xa67b3e) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #12 0x558a42a068a7  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x7528a7) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #13 0x558a4335783e  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x10a383e) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #14 0x558a4356501a  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x12b101a) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #15 0x558a42fc720a  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0xd1320a) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #16 0x558a429ae06c  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x6fa06c) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #17 0x558a42a5b83d  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x7a783d) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #18 0x558a42b05a78  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x851a78) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #19 0x558a437c7cca  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1513cca) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #20 0x558a43519b30  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1265b30) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #21 0x558a43470edb  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x11bcedb) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #22 0x558a43471f07  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x11bdf07) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #23 0x558a434759c0  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x11c19c0) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #24 0x558a43473bc9  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x11bfbc9) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #25 0x558a438111ad  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x155d1ad) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #26 0x558a432c2120  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x100e120) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #27 0x558a432bfa99  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x100ba99) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #28 0x558a4347301f  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x11bf01f) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #29 0x558a43473ede  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x11bfede) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)

Thread T19 created by T2 here:
    #0 0x558a42913e51  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x65fe51) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #1 0x558a4413f711  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1e8b711) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #2 0x558a43f93867  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1cdf867) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #3 0x558a43f92969  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1cde969) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #4 0x558a43f94cad  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1ce0cad) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #5 0x558a43f22030  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1c6e030) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #6 0x558a43c48f96  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1994f96) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #7 0x558a43c46efd  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1992efd) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #8 0x558a43c0162f  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x194d62f) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #9 0x558a43c055bb  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x19515bb) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #10 0x558a43c9b0df  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x19e70df) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #11 0x558a43512042  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x125e042) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #12 0x558a436fa916  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1446916) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #13 0x558a436f96d9  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x14456d9) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #14 0x558a42d5f5b2  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0xaab5b2) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #15 0x558a42d2bd81  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0xa77d81) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #16 0x558a42a29f32  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x775f32) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #17 0x558a43353bfa  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x109fbfa) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #18 0x558a4356eb50  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x12bab50) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #19 0x558a42fc720a  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0xd1320a) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #20 0x558a429a5fa5  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x6f1fa5) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #21 0x558a42a64e4d  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x7b0e4d) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #22 0x558a42afee78  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x84ae78) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #23 0x558a437c906a  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x151506a) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #24 0x558a43519b30  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1265b30) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #25 0x558a43470edb  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x11bcedb) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #26 0x558a43471f07  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x11bdf07) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #27 0x558a434759c0  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x11c19c0) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #28 0x558a43473bc9  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x11bfbc9) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #29 0x558a438111ad  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x155d1ad) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #30 0x558a432c2120  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x100e120) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #31 0x558a432bfa99  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x100ba99) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #32 0x558a4347301f  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x11bf01f) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #33 0x558a43473ede  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x11bfede) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #34 0x558a4343c8ff  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x11888ff) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #35 0x558a4371a865  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1466865) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #36 0x558a4343c427  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1188427) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #37 0x558a43811799  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x155d799) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #38 0x558a43811e68  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x155de68) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #39 0x558a43030eba  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0xd7ceba) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #40 0x558a43030b76  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0xd7cb76) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #41 0x558a4386cbc7  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x15b8bc7) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #42 0x558a43830eb3  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x157ceb3) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #43 0x558a43834601  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1580601) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #44 0x558a4413f8da  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1e8b8da) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #45 0x558a4292c176  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x678176) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)

Thread T2 created by T0 here:
    #0 0x558a42913e51  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x65fe51) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #1 0x558a4413f711  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1e8b711) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #2 0x558a4386abd0  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x15b6bd0) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #3 0x558a4384b117  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1597117) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #4 0x558a43867c1b  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x15b3c1b) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #5 0x558a43868aca  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x15b4aca) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #6 0x558a433df5f2  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x112b5f2) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #7 0x558a4412c55f  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x1e7855f) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #8 0x558a43078b08  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0xdc4b08) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #9 0x558a433df61d  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x112b61d) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)
    #10 0x7f5525e2a28a  (/lib/x86_64-linux-gnu/libc.so.6+0x2a28a) (BuildId: 282c2c16e7b6600b0b22ea0c99010d2795752b5f)
    #11 0x558a428a60a4  (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x5f20a4) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0)

SUMMARY: AddressSanitizer: heap-use-after-free (/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30+0x60bb5d) (BuildId: a1bf35a058700f1e22c6beaef0f94b8d3b8617e0) 
Shadow bytes around the buggy address:
  0x5250000bed80: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x5250000bee00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x5250000bee80: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x5250000bef00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x5250000bef80: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x5250000bf000:[fd]fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x5250000bf080: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x5250000bf100: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x5250000bf180: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x5250000bf200: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x5250000bf280: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==10022==ABORTING
error: test failed, to rerun pass `-p foyer --lib`

Caused by:
  process didn't exit successfully: `/home/runner/work/foyer/foyer/target/x86_64-unknown-linux-gnu/debug/deps/foyer-e0932dfd3ff48a30 --nocapture` (exit status: 1)
2025-09-16T07:58:26.302301Z  INFO foyer_storage::store: [store]: Dedicated runtime is disabled. This may lead to spikes in latency under high load. Hint: Consider configuring a dedicated runtime.
Error: Process completed with exit code 1.
```

--- a/foyer-common/src/properties.rs
+++ b/foyer-common/src/properties.rs
@@ -47,24 +47,19 @@ impl Default for Hint {
 ///
 /// NOTE: `CacheLocation` only affects the first time the entry is handle.
 /// After it is populated, the entry may not follow the given advice.
-#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
 #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
 pub enum Location {
     /// The default location.
     ///
     /// Prefer to store the entry in the in-memory cache with in-memory cache.
     /// And prefer to store the entry in the hybrid cache with hybrid cache.
+    #[default]
     Default,
     /// Prefer to store the entry in the in-memory cache.
     InMem,
     /// Prefer to store the entry on the disk cache.
     OnDisk,
-}
-
-impl Default for Location {
-    fn default() -> Self {
-        Self::Default
-    }
 }
 
 /// Entry age in the disk cache. Used by hybrid cache.
@@ -106,16 +101,16 @@ impl Default for Source {
 /// The in-memory only cache and the hybrid cache may have different properties implementations to minimize the overhead
 /// of necessary properties in different scenarios.
 pub trait Properties: Send + Sync + 'static + Clone + Default + Debug {
-    /// Set disposable.
+    /// Set entry as a phantom entry.
     ///
-    /// If an entry is disposable, it will not actually inserted into the cache and removed immediately after the last
-    /// reference drops.
+    /// A phantom entry will not be actually inserted into the in-memory cache.
+    /// It is only used to keep the APIs consistent.
     ///
-    /// Disposable property is used to simplify the design consistency of the APIs.
-    fn with_disposable(self, disposable: bool) -> Self;
+    /// NOTE: This API is for internal usage only. It MUST NOT be exported publicly.
+    fn with_phantom(self, phantom: bool) -> Self;
 
-    /// Entry disposable.
-    fn disposable(&self) -> Option<bool>;
+    /// If the entry is a phantom entry.
+    fn phantom(&self) -> Option<bool>;
 
     /// Set entry hint.
     fn with_hint(self, hint: Hint) -> Self;

--- a/foyer-memory/src/cache.rs
+++ b/foyer-memory/src/cache.rs
@@ -43,26 +43,19 @@ use crate::{
 /// Entry properties for in-memory only cache.
 #[derive(Debug, Clone, Default)]
 pub struct CacheProperties {
-    disposable: bool,
-
+    phantom: bool,
     hint: Hint,
 }
 
 impl CacheProperties {
     /// Set disposable.
-    ///
-    /// If an entry is disposable, it will not actually inserted into the cache and removed immediately after the last
-    /// reference drops.
-    ///
-    /// Disposable property is used to simplify the design consistency of the APIs.
-    pub fn with_disposable(mut self, disposable: bool) -> Self {
-        self.disposable = disposable;
+    fn with_phantom(mut self, phantom: bool) -> Self {
+        self.phantom = phantom;
         self
     }
 
-    /// Get if the entry is disposable.
-    pub fn disposable(&self) -> bool {
-        self.disposable
+    fn phantom(&self) -> bool {
+        self.phantom
     }
 
     /// Set entry hint.
@@ -78,12 +71,12 @@ impl CacheProperties {
 }
 
 impl Properties for CacheProperties {
-    fn with_disposable(self, disposable: bool) -> Self {
-        self.with_disposable(disposable)
+    fn with_phantom(self, phantom: bool) -> Self {
+        self.with_phantom(phantom)
     }
 
-    fn disposable(&self) -> Option<bool> {
-        Some(self.disposable())
+    fn phantom(&self) -> Option<bool> {
+        Some(self.phantom())
     }
 
     fn with_hint(self, hint: Hint) -> Self {

--- a/foyer-memory/src/eviction/test_utils.rs
+++ b/foyer-memory/src/eviction/test_utils.rs
@@ -81,20 +81,20 @@ pub fn assert_ptr_vec_vec_eq<T>(vva: Vec<Vec<Arc<T>>>, vvb: Vec<Vec<Arc<T>>>) {
 /// Properties for test, support all properties.
 #[derive(Debug, Clone, Default)]
 pub struct TestProperties {
-    disposable: bool,
+    phantom: bool,
     hint: Hint,
     location: Location,
     source: Source,
 }
 
 impl Properties for TestProperties {
-    fn with_disposable(mut self, disposable: bool) -> Self {
-        self.disposable = disposable;
+    fn with_phantom(mut self, phantom: bool) -> Self {
+        self.phantom = phantom;
         self
     }
 
-    fn disposable(&self) -> Option<bool> {
-        Some(self.disposable)
+    fn phantom(&self) -> Option<bool> {
+        Some(self.phantom)
     }
 
     fn with_hint(mut self, hint: Hint) -> Self {

--- a/foyer-memory/src/raw.rs
+++ b/foyer-memory/src/raw.rs
@@ -13,7 +13,7 @@
 // limitations under the License.
 
 use std::{
-    collections::hash_map::{Entry as HashMapEntry, HashMap},
+    // collections::hash_map::{Entry as HashMapEntry, HashMap},
     fmt::Debug,
     future::Future,
     hash::Hash,
@@ -40,6 +40,7 @@ use foyer_common::{
     strict_assert,
     utils::scope::Scope,
 };
+use hashbrown::hash_map::{Entry as HashMapEntry, HashMap};
 use itertools::Itertools;
 use parking_lot::{Mutex, RwLock};
 use pin_project::pin_project;
@@ -140,8 +141,27 @@ where
         record: Arc<Record<E>>,
         garbages: &mut Vec<(Event, Arc<Record<E>>)>,
         waiters: &mut Vec<oneshot::Sender<RawCacheEntry<E, S, I>>>,
-    ) -> Arc<Record<E>> {
+    ) {
         *waiters = self.waiters.lock().remove(record.key()).unwrap_or_default();
+
+        if record.properties().phantom().unwrap_or_default() {
+            if let Some(old) = self.indexer.remove(record.hash(), record.key()) {
+                strict_assert!(!old.is_in_indexer());
+
+                if old.is_in_eviction() {
+                    self.eviction.remove(&old);
+                }
+                strict_assert!(!old.is_in_eviction());
+
+                self.usage -= old.weight();
+
+                garbages.push((Event::Replace, old));
+            }
+            record.inc_refs(waiters.len() + 1);
+            garbages.push((Event::Remove, record));
+            self.metrics.memory_insert.increase(1);
+            return;
+        }
 
         let weight = record.weight();
         let old_usage = self.usage;
@@ -182,8 +202,6 @@ where
             std::cmp::Ordering::Less => self.metrics.memory_usage.decrease((old_usage - self.usage) as _),
             std::cmp::Ordering::Equal => {}
         }
-
-        record
     }
 
     #[cfg_attr(feature = "tracing", fastrace::trace(name = "foyer::memory::raw::shard::remove"))]
@@ -575,7 +593,12 @@ where
         let hash = self.inner.hash_builder.hash_one(&key);
         let weight = (self.inner.weighter)(&key, &value);
         if !(self.inner.filter)(&key, &value) {
-            properties = properties.with_disposable(true);
+            properties = properties.with_phantom(true);
+        }
+        if let Some(location) = properties.location() {
+            if location == Location::OnDisk {
+                properties = properties.with_phantom(true);
+            }
         }
         let record = Arc::new(Record::new(Data {
             key,
@@ -595,38 +618,12 @@ where
 
     #[cfg_attr(feature = "tracing", fastrace::trace(name = "foyer::memory::raw::insert_inner"))]
     fn insert_inner(&self, record: Arc<Record<E>>) -> RawCacheEntry<E, S, I> {
-        if record.properties().disposable().unwrap_or_default() {
-            // Remove the stale record if it exists.
-            self.inner.shards[self.shard(record.hash())]
-                .write()
-                .remove(record.hash(), record.key())
-                .inspect(|r| {
-                    // Deallocate data out of the lock critical section.
-                    if let Some(listener) = self.inner.event_listener.as_ref() {
-                        listener.on_leave(Event::Evict, r.key(), r.value());
-                    }
-                });
-
-            let pipe = self.inner.pipe.load();
-            if pipe.is_enabled() {
-                pipe.send(Piece::new(record.clone()));
-            }
-
-            // If the record is disposable, we do not insert it into the cache.
-            // Instead, we just return it and let it be dropped immediately after the last reference drops.
-            record.inc_refs(1);
-            return RawCacheEntry {
-                record,
-                inner: self.inner.clone(),
-            };
-        }
-
         let mut garbages = vec![];
         let mut waiters = vec![];
 
-        let record = self.inner.shards[self.shard(record.hash())]
+        self.inner.shards[self.shard(record.hash())]
             .write()
-            .with(|mut shard| shard.emplace(record, &mut garbages, &mut waiters));
+            .with(|mut shard| shard.emplace(record.clone(), &mut garbages, &mut waiters));
 
         // Notify waiters out of the lock critical section.
         for waiter in waiters {
@@ -848,8 +845,14 @@ where
         let shard = &self.inner.shards[hash as usize % self.inner.shards.len()];
 
         if self.record.dec_refs(1) == 0 {
-            if self.record.properties().disposable().unwrap_or_default() {
-                // TODO(MrCroxx): Send it to disk cache write queue with pipe?
+            if self.record.properties().phantom().unwrap_or_default() {
+                if let Some(listener) = self.inner.event_listener.as_ref() {
+                    listener.on_leave(Event::Evict, self.record.key(), self.record.value());
+                }
+                let pipe = self.inner.pipe.load();
+                if pipe.is_enabled() {
+                    pipe.send(Piece::new(self.record.clone()));
+                }
                 return;
             }
 
@@ -1127,18 +1130,25 @@ where
 
         match raw {
             RawShardFetch::Hit(record) => {
+                tracing::trace!(hash, "fetch => Hit");
                 return RawFetch::new(RawFetchInner::Hit(Some(RawCacheEntry {
                     record,
                     inner: self.inner.clone(),
-                })))
+                })));
             }
-            RawShardFetch::Wait(future) => return RawFetch::new(RawFetchInner::Wait(future)),
-            RawShardFetch::Miss => {}
+            RawShardFetch::Wait(future) => {
+                tracing::trace!(hash, "fetch => Wait");
+                return RawFetch::new(RawFetchInner::Wait(future));
+            }
+            RawShardFetch::Miss => {
+                tracing::trace!(hash, "fetch => Miss");
+            }
         }
 
         let cache = self.clone();
         let future = fetch();
         let join = runtime.spawn({
+            tracing::trace!(hash, "fetch => join !!!");
             let task = async move {
                 #[cfg(feature = "tracing")]
                 let Diversion { target, store } = future
@@ -1158,12 +1168,13 @@ where
                 };
                 if let Some(ctx) = store.as_ref() {
                     if ctx.throttled {
+                        // TODO(MrCroxx): Make sure foyer doesn't issue tombstone to the disk cache.
+                        // TODO(MrCroxx): Also make sure the disk cache will write tombstone instead of this case.
                         properties = properties.with_location(Location::InMem)
                     }
                     properties = properties.with_source(ctx.source)
                 };
-                let location = properties.location().unwrap_or_default();
-                properties = properties.with_disposable(matches! {location, Location::OnDisk});
+                tracing::trace!(hash, "fetch => insert !!!");
                 let entry = match target.into() {
                     FetchTarget::Value(value) => cache.insert_with_properties(key, value, properties),
                     FetchTarget::Piece(p) => cache.insert_inner(p.into_record::<E>()),
@@ -1289,15 +1300,15 @@ mod tests {
     }
 
     #[test_log::test]
-    fn test_insert_disposable() {
+    fn test_insert_phantom() {
         let fifo = fifo_cache_for_test();
 
-        let e1 = fifo.insert_with_properties(1, 1, TestProperties::default().with_disposable(true));
+        let e1 = fifo.insert_with_properties(1, 1, TestProperties::default().with_phantom(true));
         assert_eq!(fifo.usage(), 0);
         drop(e1);
         assert_eq!(fifo.usage(), 0);
 
-        let e2a = fifo.insert_with_properties(2, 2, TestProperties::default().with_disposable(true));
+        let e2a = fifo.insert_with_properties(2, 2, TestProperties::default().with_phantom(true));
         assert_eq!(fifo.usage(), 0);
         assert!(fifo.get(&2).is_none());
         assert_eq!(fifo.usage(), 0);
@@ -1308,7 +1319,7 @@ mod tests {
         fifo.insert(1, 1);
         assert_eq!(fifo.usage(), 1);
         assert_eq!(fifo.get(&1).unwrap().value(), &1);
-        let e2 = fifo.insert_with_properties(1, 100, TestProperties::default().with_disposable(true));
+        let e2 = fifo.insert_with_properties(1, 100, TestProperties::default().with_phantom(true));
         assert_eq!(fifo.usage(), 0);
         drop(e2);
         assert_eq!(fifo.usage(), 0);

--- a/foyer/Cargo.toml
+++ b/foyer/Cargo.toml
@@ -18,15 +18,16 @@ rustdoc-args = ["--cfg", "docsrs"]
 
 [features]
 default = []
-serde = ["foyer-common/serde", "foyer-storage/serde"]
 clap = ["foyer-storage/clap"]
+serde = ["foyer-common/serde", "foyer-storage/serde"]
+nightly = ["foyer-storage/nightly", "foyer-memory/nightly"]
 tracing = [
   "dep:fastrace",
   "foyer-common/tracing",
   "foyer-memory/tracing",
   "foyer-storage/tracing",
 ]
-nightly = ["foyer-storage/nightly", "foyer-memory/nightly"]
+test_utils = ["foyer-storage/test_utils"]
 deadlock = ["foyer-storage/deadlock"]
 strict_assertions = [
   "foyer-common/strict_assertions",
@@ -69,9 +70,15 @@ tokio = { workspace = true, features = [
 ] }
 
 [dev-dependencies]
+foyer = { workspace = true, features = ["test_utils"] }
 foyer-storage = { workspace = true, features = ["test_utils"] }
+rand = "0.9"
 tempfile = { workspace = true }
 test-log = { workspace = true, features = ["trace", "color"] }
 
 [lints]
 workspace = true
+
+[[test]]
+name = "hybrid_cache_fuzzy_test"
+required-features = ["test_utils"]

--- a/foyer/src/hybrid/cache.rs
+++ b/foyer/src/hybrid/cache.rs
@@ -93,27 +93,20 @@ macro_rules! try_cancel {
 /// Entry properties for in-memory only cache.
 #[derive(Debug, Clone, Default)]
 pub struct HybridCacheProperties {
-    disposable: bool,
+    phantom: bool,
     hint: Hint,
     location: Location,
     source: Source,
 }
 
 impl HybridCacheProperties {
-    /// Set disposable.
-    ///
-    /// If an entry is disposable, it will not actually inserted into the cache and removed immediately after the last
-    /// reference drops.
-    ///
-    /// Disposable property is used to simplify the design consistency of the APIs.
-    fn with_disposable(mut self, disposable: bool) -> Self {
-        self.disposable = disposable;
+    fn with_phantom(mut self, phantom: bool) -> Self {
+        self.phantom = phantom;
         self
     }
 
-    /// Get if the entry is disposable.
-    fn disposable(&self) -> bool {
-        self.disposable
+    fn phantom(&self) -> bool {
+        self.phantom
     }
 
     /// Set entry hint.
@@ -145,12 +138,12 @@ impl HybridCacheProperties {
 }
 
 impl Properties for HybridCacheProperties {
-    fn with_disposable(self, disposable: bool) -> Self {
-        self.with_disposable(disposable)
+    fn with_phantom(self, phantom: bool) -> Self {
+        self.with_phantom(phantom)
     }
 
-    fn disposable(&self) -> Option<bool> {
-        Some(self.disposable())
+    fn phantom(&self) -> Option<bool> {
+        Some(self.phantom())
     }
 
     fn with_hint(self, hint: Hint) -> Self {
@@ -565,11 +558,7 @@ where
 
         let now = Instant::now();
 
-        let disposable = matches! { properties.location(), Location::OnDisk };
-        let entry = self
-            .inner
-            .memory
-            .insert_with_properties(key, value, properties.with_disposable(disposable));
+        let entry = self.inner.memory.insert_with_properties(key, value, properties);
         if self.inner.policy == HybridCachePolicy::WriteOnInsertion && entry.properties().location() != Location::InMem
         {
             self.inner.storage.enqueue(entry.piece(), false);

--- a/foyer/src/hybrid/writer.rs
+++ b/foyer/src/hybrid/writer.rs
@@ -23,7 +23,7 @@ use foyer_common::{
 };
 use foyer_storage::StorageFilterResult;
 
-use crate::{HybridCache, HybridCacheEntry, HybridCachePolicy, HybridCacheProperties};
+use crate::{HybridCache, HybridCacheEntry, HybridCacheProperties};
 
 /// Writer for hybrid cache to support more flexible write APIs.
 pub struct HybridCacheWriter<K, V, S = DefaultHasher>
@@ -141,11 +141,7 @@ where
 
         let entry = self
             .hybrid
-            .memory()
-            .insert_with_properties(self.key, value, properties.with_disposable(true));
-        if self.hybrid.policy() == HybridCachePolicy::WriteOnInsertion {
-            self.hybrid.storage().enqueue(entry.piece(), true);
-        }
+            .insert_with_properties(self.key, value, properties.with_phantom(true));
         self.hybrid.metrics().hybrid_insert.increase(1);
         self.hybrid
             .metrics()

--- a/foyer/tests/hybrid_cache_fuzzy_test.rs
+++ b/foyer/tests/hybrid_cache_fuzzy_test.rs
@@ -1,0 +1,177 @@
+// Copyright 2025 foyer Project Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Fuzzy test for foyer hybrid cache.
+
+use std::{
+    collections::VecDeque,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc, RwLock,
+    },
+    time::Duration,
+};
+
+use foyer::{
+    BlockEngineBuilder, DeviceBuilder, Event, EventListener, FsDeviceBuilder, HybridCache, HybridCacheBuilder,
+    HybridCachePolicy, HybridCacheProperties, IoEngineBuilder, Location, PsyncIoEngineBuilder,
+};
+use rand::{rng, Rng};
+
+const KB: usize = 1024;
+const MB: usize = 1024 * KB;
+
+const WRITERS: usize = 8;
+const READERS: usize = 16;
+
+const WRITES: usize = 1000;
+const DUPLICATES: usize = 10;
+const READS: usize = 1000;
+
+const FETCH_WAIT: Duration = Duration::from_millis(10);
+const WRITE_WAIT: Duration = Duration::from_millis(1);
+const READ_WAIT: Duration = Duration::from_millis(1);
+
+const INTERVAL: usize = 100;
+
+#[derive(Debug)]
+struct RecentEvictionQueue {
+    queue: RwLock<VecDeque<u64>>,
+    capacity: usize,
+}
+
+impl EventListener for RecentEvictionQueue {
+    type Key = u64;
+    type Value = Vec<u8>;
+
+    fn on_leave(&self, _: Event, key: &Self::Key, _: &Self::Value) {
+        let mut queue = self.queue.write().unwrap();
+        if queue.len() < self.capacity {
+            queue.push_back(*key)
+        } else {
+            queue.pop_front();
+            queue.push_back(*key);
+        }
+    }
+}
+
+impl RecentEvictionQueue {
+    fn new(capacity: usize) -> Self {
+        Self {
+            queue: RwLock::new(VecDeque::with_capacity(capacity)),
+            capacity,
+        }
+    }
+
+    fn pick(&self) -> Option<u64> {
+        let queue = self.queue.read().unwrap();
+        if queue.is_empty() {
+            return None;
+        }
+        let index = rng().random_range(0..queue.len());
+        Some(queue[index])
+    }
+}
+
+#[test_log::test(tokio::test)]
+async fn test_concurrent_insert_disk_cache_and_fetch() {
+    let dir = tempfile::tempdir().unwrap();
+
+    let recent = Arc::new(RecentEvictionQueue::new(10));
+
+    let hybrid: HybridCache<u64, Vec<u8>> = HybridCacheBuilder::new()
+        .with_name("test")
+        .with_policy(HybridCachePolicy::WriteOnEviction)
+        .with_event_listener(recent.clone())
+        .memory(MB)
+        .with_weighter(|_, v| 8 + v.len())
+        .storage()
+        .with_io_engine(PsyncIoEngineBuilder::new().build().await.unwrap())
+        .with_engine_config(
+            BlockEngineBuilder::new(FsDeviceBuilder::new(dir).with_capacity(64 * MB).build().unwrap())
+                .with_block_size(4 * MB),
+        )
+        .build()
+        .await
+        .unwrap();
+
+    let idx = Arc::new(AtomicU64::new(0));
+
+    let mut handles = vec![];
+    for _ in 0..WRITERS {
+        let h = hybrid.clone();
+        let r = recent.clone();
+        let i = idx.clone();
+        handles.push(tokio::spawn(async move { write(h, r, i).await }));
+    }
+    for _ in 0..READERS {
+        let h = hybrid.clone();
+        let r = recent.clone();
+        handles.push(tokio::spawn(async move { read(h, r).await }));
+    }
+    for h in handles {
+        h.await.unwrap();
+    }
+}
+
+fn value(key: u64) -> Vec<u8> {
+    vec![key as u8; 4 * KB]
+}
+
+async fn write(hybrid: HybridCache<u64, Vec<u8>>, _: Arc<RecentEvictionQueue>, idx: Arc<AtomicU64>) {
+    loop {
+        let key = idx.fetch_add(1, Ordering::Relaxed);
+        if key > WRITES as u64 {
+            break;
+        }
+        if key % INTERVAL as u64 == 0 {
+            tracing::info!("Inserted {key} items");
+        }
+        for k in key.saturating_sub(DUPLICATES as u64)..=key {
+            hybrid.insert_with_properties(
+                k,
+                value(k),
+                HybridCacheProperties::default().with_location(Location::OnDisk),
+            );
+            tokio::time::sleep(WRITE_WAIT).await;
+        }
+    }
+}
+
+async fn read(hybrid: HybridCache<u64, Vec<u8>>, recent: Arc<RecentEvictionQueue>) {
+    let mut cnt = 0;
+    loop {
+        tokio::time::sleep(READ_WAIT).await;
+        let key = match recent.pick() {
+            Some(v) => v,
+            None => continue,
+        };
+        let e = hybrid
+            .fetch(key, || async move {
+                tokio::time::sleep(FETCH_WAIT).await;
+                Ok(value(key))
+            })
+            .await
+            .unwrap();
+
+        assert_eq!(e.value(), &value(key));
+        cnt += 1;
+        if cnt % INTERVAL as u64 == 0 {
+            tracing::info!("Read {cnt} items");
+        }
+        if cnt >= READS as u64 {
+            break;
+        }
+    }
+}

